src/StorageWriter/ClaimController.ts

```diff
@@ -1,51 +1,102 @@
   private readonly logger: Pino.Logger
-  private readonly db: Db
-  private readonly collection: Collection
-  private readonly messaging: Messaging
+  private readonly db: Database
   private readonly ipfs: IPFS
 
-  constructor(
-    @inject('Logger') logger: Pino.Logger,
-    @inject('DB') db: Db,
-    @inject('Messaging') messaging: Messaging,
-    @inject('IPFS') ipfs: IPFS
-  ) {
+  constructor(@inject('Logger') logger: Pino.Logger, @inject('Database') db: Database, @inject('IPFS') ipfs: IPFS) {
     this.logger = childWithFileName(logger, __filename)
     this.db = db
-    this.collection = this.db.collection('storageWriter')
-    this.messaging = messaging
     this.ipfs = ipfs
   }
 
-  async create(claim: Claim): Promise<void> {
-    const logger = this.logger.child({ method: 'create' })
+  private readonly log = (level: LogTypes) => (message: string) => async (value: any) => {
+    const logger = this.logger
+    logger[level]({ value }, message)
+    return value
+  }
 
-    logger.trace({ claim }, 'Storing Claim')
+  // tslint:disable-next-line
+  public readonly addClaim = async (claim: Claim): Promise<void> => {
+    const logger = this.logger.child({ method: 'addClaim' })
 
-    const ipfsFileHash = await this.ipfs.addText(JSON.stringify(claim))
+    logger.trace({ claim }, 'Adding Claim')
 
-    logger.info({ claim, ipfsFileHash }, 'Claim Stored')
+    await this.db.addClaim(claim)
 
-    await this.collection.insertOne({
-      claimId: claim.id,
-      ipfsFileHash,
-    })
-    await this.messaging.publish(Exchange.ClaimIPFSHash, {
-      claimId: claim.id,
-      ipfsFileHash,
-    })
+    logger.info({ claim }, 'Claim Added')
   }
+
+  private readonly findNextClaim = async (): Promise<StoreNextClaimData> => {
+    const claim = await this.db.findNextClaim()
+    return { claim }
+  }
+
+  private readonly uploadClaim = (claim: Claim) => this.ipfs.addText(JSON.stringify(claim))
+
+  private readonly handleStoreClaimError = async (error: Error, claim: Claim) => {
+    await this.db.addError(uploadClaimError(error.message, claim))
+    throw new Error('Failed to store claim')
+  }
+
+  private readonly storeClaim = async (data: StoreNextClaimData): Promise<StoreNextClaimData> => {
+    const { claim } = data
+    try {
+      const ipfsFileHash = await this.uploadClaim(claim)
+      return {
+        ...data,
+        ipfsFileHash,
+      }
+    } catch (error) {
+      await this.handleStoreClaimError(error, claim)
+    }
+  }
+
+  private readonly handleAddIPFSHashToClaimError = async (error: Error, claim: Claim, hash: string) => {
+    await this.db.addError(updateClaimHashError(error.message, claim, hash))
+    throw new Error('Failed to update claim hash')
+  }
+
+  private readonly addIPFSHashToClaim = async (data: StoreNextClaimData): Promise<StoreNextClaimData> => {
+    const { claim, ipfsFileHash } = data
+    try {
+      await this.db.addClaimHash(claim.id, ipfsFileHash)
+      return data
+    } catch (error) {
```

Comment:
This is a good point. I was thinking about this wrong. I was trying to use Mongo for persisted errors because I don't think we are persisting our logging to a file or anything at the moment are we?

Reply:
Log persistence is not a responsibility of the application emitting the logs, but it WILL be done.

Applications should just write to stdout. When running in a Docker container, Docker will grab all the stdout and store it in a file for you. Another application can grab those logs and send them somewhere else.

Logz.io will handle everything for us for now, including persistence. Elastic Stack in the future, maybe.

So yeah, if that was the main reason to store them in the DB, then there's no need.